File: packages/snippets-ts/src/apply-snippets.ts
import type { Config } from './config'

interface Match {
  readonly body: string
  readonly length: number
}

function findMatch(input: string, cursor: number, config: Config): Match | null {
  for (const snippet of config.snippets) {
    for (const matcher of snippet.matchers) {
      matcher.regex.lastIndex = cursor

      const result = matcher.regex.exec(input)

      if (result === null || result[0].length === 0) {
        continue
      }

      return { body: snippet.body, length: result[0].length }
    }
  }

  return null
}

export function applySnippets(input: string, config: Config): string {
  let output = ''
  let cursor = 0

  while (cursor < input.length) {
    const match = findMatch(input, cursor, config)

    if (match !== null) {
      output += match.body
      cursor += match.length
      continue
    }

    const codePoint = input.codePointAt(cursor)

    if (codePoint === undefined) {
      throw new TypeError('cursor moved outside the input')
    }

    const character = String.fromCodePoint(codePoint)

    output += character
    cursor += character.length
  }

  return output
}

File: packages/snippets-rs/src/apply_snippets.rs
use crate::parse_config::Config;

fn find_match<'a>(input: &str, cursor: usize, config: &'a Config) -> Option<(&'a str, usize)> {
    for snippet in &config.snippets {
        for matcher in &snippet.matchers {
            let Some(found) = matcher.regex.find_from(input, cursor).next() else {
                continue;
            };
            let range = found.range();

            if range.start != cursor || range.end == range.start {
                continue;
            }

            return Some((snippet.body.as_str(), range.end - range.start));
        }
    }

    None
}

pub fn apply_snippets(input: &str, config: &Config) -> String {
    let mut output = String::new();
    let mut cursor = 0;

    while cursor < input.len() {
        if let Some((body, length)) = find_match(input, cursor, config) {
            output.push_str(body);
            cursor += length;

            continue;
        }

        let character = input[cursor..]
            .chars()
            .next()
            .expect("cursor must rest on a character boundary inside the input");

        output.push(character);
        cursor += character.len_utf8();
    }

    output
}
